### ISSUE TITLE
Metrics "fastlookup" is Using RubyArray as Hash Map Key

### ISSUE THREAD

**Issue by original-brownbear** [elastic/logstash#7772]
I think this is one of if not the main issue behind #7396 

Look at the profile of just running `-e 'input { stdin{}} output {stdout {codec => "dots"}}'` with some large file piped to standard input.

<img width="1809" alt="screen shot 2017-07-23 at 10 28 00" src="https://user-images.githubusercontent.com/6490959/28497873-a0dea730-6f91-11e7-9f56-d99821605681.png">

We're wasting an absurd amount of time looking up from the `@fast_lookup` in `metric_store.rb`.

The reason for this is trivial, this is what we use for map keys:

`namespaces.dup << key`

`namespaces` is an array and hence the whole hashmap key is an array.
This wouldn't even work with a Java hash, but (un)fortunately JRuby's `RubyArray` uses this method as the override for `equals()`:

```rb
    /** rb_ary_equal
     *
     */
    @JRubyMethod(name = "==", required = 1)
    @Override
    public IRubyObject op_equal(ThreadContext context, IRubyObject obj) {
        Ruby runtime = context.runtime;

        if (this == obj) {
            return runtime.getTrue();
        }

        if (!(obj instanceof RubyArray)) {
            if (obj == context.nil) return runtime.getFalse();

            if (!sites(context).respond_to_to_ary.respondsTo(context, obj, obj)) {
                return runtime.getFalse();
            }
            return Helpers.rbEqual(context, obj, this);
        }
        return RecursiveComparator.compare(context, sites(context).op_equal, this, obj);
    }
```

... which seems to be absurdly expensive (kind of expected, it eventually just compares all array elements).
Also literally every call to this thing requires creating at least one new `RubyArray` instance!

The solution here is obviously to:

* Use an immutable hash key with a cheap and evenly distributed hashcode as well as a fast `equals()` method (like `java.lang.String`)
* Use a key approach that doesn't require creating new objects on lookup

**Comment by original-brownbear** [comment]
@jakelandis @andrewvc I have on (admittedly extremely annoying :)) question in all of this:

Why don't we simply port our metric system to DW metrics (http://metrics.dropwizard.io/3.2.3/) instead of trying to solve issues like this one ourselves when DW has this figured out already (we could also save some other code we have around monitoring the JVM by reuse from that lib).

Just my 2 cents on this, given that this one looks like a somewhat hard problem :)

**Comment by andrewvc** [comment]
@original-brownbear we already have a plan to move away from generic metrics libraries to simple POJOs with our new `Gauge` and `Counter` classes. After assessing our needs we determined we don't need anything more complex than that. That move should fix this whole lookup issue. We won't need any hashmaps, caching or other expensive stuff once we make that move.

WRT monitoring JVM usage, we already have the code and it works. I don't see us needing to modify it going forward, so I don't see a concrete reason to refactor.

**Comment by andrewvc** [comment]
@jakelandis do you know if we have a formal issue for this refactor yet? We have #7155 , but we don't yet have one for the POJO-ification we've discussed. I'm glad to write one up

**Comment by jakelandis** [comment]
@andrewvc - Not explicitly issue for POJO-fication , just a PR https://github.com/elastic/logstash/pull/7274 that starts us down that path and a discussion to start depreciating metrics https://github.com/elastic/logstash/issues/7746

**Comment by original-brownbear** [comment]
@andrewvc @jakelandis sorry never saw that PR ... We're much further along than I expected :)

Nevermind then. I guess The move to POJO's will automatically fix this issue then anyways :)

**Comment by andrewvc** [comment]
@original-brownbear no worries! It's my bad for not making a real design issue. I've just written one up here: https://github.com/elastic/logstash/issues/7788

**Comment by jakelandis** [comment]
@original-brownbear - Nice find!  I was able to see the hot method via YourKit.  I tried a couple different approaches, but short of a re-write, I think that using the RubyArray.hash is our best bet moving forward. 

The RubyArray.hash still requires, upon every lookup, calculating the hash from the values of the array, but saves time by not requiring walking the array during the equals check. From both YourKit and wall time of 10M events, it is an improvement.  ~10-15% speed up on a pipeline with apache logs input, no filters and null output. 

The primary consumption of this will be replaced in 6.1 via #7788 and will be completely gone by 7.0.0.  

I would suggest that we push this fix to the 6.0.0-beta release, maybe a 5.x release ? @suyograo ?

Before w/ 10M apache web server logs, no filters, null output:
```
real	6m28.988s
user	8m36.299s
sys	0m46.656s
```
After:
```
real	5m44.258s
user	7m48.024s
sys	0m43.809s
```
